### ISSUE TITLE
Memory leak when blitting on MacOS/M1 

### ISSUE THREAD

**Issue by colormotor** [pyglet/pyglet#1147]
Hi, I am running pyglet on MacOS Ventura, M1 processor and Python 3.10. 
I am noticing a significant memory leak when using `blit` with either an image or `ImageData`. 
This code reproduces the problem on my machine:
```Python
import pyglet

image = pyglet.image.load('test.jpg')
window = pyglet.window.Window(512, 512)

@window.event
def on_draw():
    window.clear()
    image.blit(0, 0)

pyglet.app.run(interval=0)
``` 
With `interval=0` memory grows approximately 10Mb every second. I tested the same on Linux and memory usage does not grow.
Any help would be greatly appreciated!

This is the output of pyglet.info on my machine:
```
Platform
------------------------------------------------------------------------------
platform:  macOS-13.6.6-arm64-arm-64bit
release:   22.6.0
machine:   arm64

Python
------------------------------------------------------------------------------
implementation: CPython
sys.version: 3.10.8 | packaged by conda-forge | (main, Nov 22 2022, 08:25:29) [Clang 14.0.6 ]
sys.maxint: 9223372036854775807
objc.__version__: 9.0
os.getcwd(): /Users/colormotor

pyglet
------------------------------------------------------------------------------
pyglet.version: 2.0.15
pyglet.compat_platform: darwin
pyglet.__file__: /Users/colormotor/opt/miniconda3/lib/python3.10/site-packages/pyglet/__init__.py
pyglet.options['audio'] = ('xaudio2', 'directsound', 'openal', 'pulse', 'silent')
pyglet.options['debug_font'] = False
pyglet.options['debug_gl'] = True
pyglet.options['debug_gl_trace'] = False
pyglet.options['debug_gl_trace_args'] = False
pyglet.options['debug_gl_shaders'] = False
pyglet.options['debug_graphics_batch'] = False
pyglet.options['debug_lib'] = False
pyglet.options['debug_media'] = False
pyglet.options['debug_texture'] = False
pyglet.options['debug_trace'] = False
pyglet.options['debug_trace_args'] = False
pyglet.options['debug_trace_depth'] = 1
pyglet.options['debug_trace_flush'] = True
pyglet.options['debug_win32'] = False
pyglet.options['debug_input'] = False
pyglet.options['debug_x11'] = False
pyglet.options['shadow_window'] = True
pyglet.options['vsync'] = None
pyglet.options['xsync'] = True
pyglet.options['xlib_fullscreen_override_redirect'] = False
pyglet.options['search_local_libs'] = True
pyglet.options['win32_gdi_font'] = False
pyglet.options['headless'] = False
pyglet.options['headless_device'] = 0
pyglet.options['win32_disable_shaping'] = False
pyglet.options['dw_legacy_naming'] = False
pyglet.options['win32_disable_xinput'] = False
pyglet.options['com_mta'] = False
pyglet.options['osx_alt_loop'] = False

pyglet.window
------------------------------------------------------------------------------
display: <pyglet.canvas.cocoa.CocoaDisplay object at 0x10525d780>
screens[0]: CocoaScreen(x=0, y=0, width=2048, height=1280)
screens[1]: CocoaScreen(x=0, y=-1080, width=1920, height=1080)
config['double_buffer'] = 1
config['stereo'] = 0
config['buffer_size'] = 32
config['aux_buffers'] = 0
config['sample_buffers'] = 0
config['samples'] = 0
config['red_size'] = 0
config['green_size'] = 0
config['blue_size'] = 0
config['alpha_size'] = 8
config['depth_size'] = 32
config['stencil_size'] = 0
config['accum_red_size'] = 0
config['accum_green_size'] = 0
config['accum_blue_size'] = 0
config['accum_alpha_size'] = 0
config['major_version'] = 3
config['minor_version'] = 2
config['forward_compatible'] = None
config['opengl_api'] = 'gl'
config['debug'] = None
context: CocoaContext(id=4372048112, share=CocoaContext(id=4381334192, share=None))

window.context._info
------------------------------------------------------------------------------
gl_info.get_version(): (4, 1)
gl_info.get_vendor(): Apple
gl_info.get_renderer(): Apple M1

pyglet.gl.glx_info
------------------------------------------------------------------------------
GLX not available.

pyglet.media
------------------------------------------------------------------------------
audio driver: <pyglet.media.drivers.openal.adaptation.OpenALDriver object at 0x105f58ca0>

pyglet.media.ffmpeg
------------------------------------------------------------------------------
FFmpeg version: 5.1.2

pyglet.media.drivers.openal
------------------------------------------------------------------------------
Library: <CDLL '/System/Library/Frameworks/OpenAL.framework/OpenAL', handle 3a823f388 at 0x105f888e0>
Version: 1.1
Extensions:
   ALC_EXT_CAPTURE
   ALC_ENUMERATION_EXT
   ALC_EXT_MAC_OSX
   ALC_EXT_ASA
   ALC_EXT_ASA_DISTORTION
   ALC_EXT_ASA_ROGER_BEEP

pyglet.input.wintab
------------------------------------------------------------------------------
WinTab not available.
```

**Comment by caffeinepills** [comment]
Thanks for the report. I tested with Windows, and don't see a memory leak, so it may be Mac specific.

However there is a bug with blitting in 2.0.15 where it will keep creating and destroying the default shader repeatedly for every blit. This should be fixed in master branch. Could you try master branch to see if that does anything?

**Comment by colormotor** [comment]
Thanks for the quick reply, using master fixes it! Do you have timeframe on when this can be updated on PyPi?

**Comment by caffeinepills** [comment]
Hopefully sometime this week, just want to make sure a few more bug fixes get added.

**Comment by benmoran56** [comment]
This is now released as part of 2.0.16, so i'll close the ticket.
